### PR TITLE
feat(raiko-lib): unify protocol instance for on chain verification

### PR DESCRIPTION
**Pull request by CeciliaZ030**

### CODE REVIEW

File: lib/src/consts.rs [hunk 1]
```diff
@@ -282,14 +290,19 @@ mod tests {
...
+                (VerifierType::RISC0, Some(Address::default())),
+            ]),
             genesis_time: 0u64,
             seconds_per_slot: 1u64,
             is_taiko: false,
         };
 
         let json = serde_json::to_string(&spec).unwrap();
         // write to a file called chain_specs.json
+
```
Comment (Brechtpd) [inline lib/src/consts.rs:305]: empty line

File: provers/sp1/driver/src/lib.rs [hunk 2]
```diff
@@ -82,9 +71,15 @@ mod test {
...
+        let proof = client
+            .prove_groth16(&pk, stdin)
+            .expect("Sp1: proving failed");
+        // client
+        //     .verify(&proof, &vk)
+        //     .expect("Sp1: verification failed");
+
+        let contract_dir = "../contract";
+        sp1_sdk::artifacts::export_solidity_groth16_verifier(contract_dir)
+            .expect("failed to export verifier");
```
Comment (Brechtpd) [inline provers/sp1/driver/src/lib.rs:83]: probably shouldn't be part of this PR?

File: lib/src/consts.rs [hunk 3]
```diff
@@ -135,6 +135,15 @@ impl Default for Eip1559Constants {
     }
 }
 
+#[repr(u8)]
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
+pub enum VerifierType {
```
Comment (Brechtpd) [inline lib/src/consts.rs:140]: Could just reuse ProofType?

**Review by Brechtpd — APPROVED** [review]
Looks good! 

`sgx_verifier_address` removed so potentially a breaking change @smtmfft.

File: lib/src/consts.rs (hunk 3 above)
Comment (CeciliaZ030) [reply lib/src/consts.rs:140]: ProofType is in host, this thing we gotta use in lib. Reusing will cause circular dep

File: lib/src/consts.rs (hunk 1 above)
Comment (CeciliaZ030) [inline lib/src/consts.rs:305]:
```suggestion
```

File: lib/src/consts.rs (hunk 3 above)
Comment (Brechtpd) [reply lib/src/consts.rs:140]: Move ProofType to lib?

Comment (CeciliaZ030) [reply lib/src/consts.rs:140]: Don't think it's possible cuz its impl contains lots of host content

File: host/config/chain_spec_list_default.json [hunk 4]
```diff
@@ -82,7 +90,11 @@
         "l2_contract": "0x1670090000000000000000000000000000010001",
         "rpc": "https://rpc.hekla.taiko.xyz",
         "beacon_rpc": null,
-        "sgx_verifier_address": "0x532efbf6d62720d0b2a2bb9d11066e8588cae6d9",
+        "verifier_address":{
+            "SGX":"0x0000000000000000000000000000000000000000",
```
Comment (smtmfft) [inline host/config/chain_spec_list_default.json:94]: why SGX address is 0x00?

**Review by smtmfft — CHANGES_REQUESTED** [review]

**Comment by smtmfft** [comment]
> Looks good!
> 
> `sgx_verifier_address` removed so potentially a breaking change @smtmfft.

it's not in API, could be little impact I think.

File: host/config/chain_spec_list_default.json (hunk 4 above)
Comment (CeciliaZ030) [reply host/config/chain_spec_list_default.json:94]: ohh because i just put dummy values, will change

File: lib/src/protocol_instance.rs [hunk 5]
```diff
@@ -22,191 +22,176 @@ pub struct ProtocolInstance {
...
-                    self.sgx_verifier_address,
-                    self.transition.clone(),
-                    new_pubkey,
-                    self.prover,
-                    self.meta_hash(),
-                )
-                    .abi_encode()
-                    .iter()
-                    .copied()
-                    .skip(32) // TRICKY: skip the first dyn flag 0x00..20.
```
[the remark this answers is not in the document]
Comment (CeciliaZ030) [reply lib/src/protocol_instance.rs:50]: Will add it back

[the remark this answers is not in the document]
Comment (smtmfft) [reply lib/src/protocol_instance.rs:50]: Yes pls, this is a tricky point, not sure why it has different output. Could be a issue from alloy side.
For us, should be ok as long as the unittest is ok. Or maybe it's already fixed, so you need to double check if you change here.

File: lib/src/protocol_instance.rs [hunk 6]
```diff
@@ -22,191 +22,177 @@ pub struct ProtocolInstance {
...
-                &kzg_settings,
-            )
+        let gas_limit: u64 = header.gas_limit.try_into().unwrap();
+        let verifier_address = input
+            .chain_spec
+            .verifier_address
+            .get(&proof_type)
+            .unwrap_or(&Some(Address::default()))
+            .clone()
             .unwrap();
```
Comment (CeciliaZ030) [inline lib/src/protocol_instance.rs:111]:
```suggestion
            .unwrap_or(&None)
            .clone()
            .unwrap_or_default();
```
